File: src/main/java/io/chesslave/eyes/Vision.java
package io.chesslave.eyes;

import javaslang.collection.Stream;
import javaslang.control.Option;
import java.awt.Rectangle;
import java.awt.image.BufferedImage;

public interface Vision {

    interface Recogniser {

        Stream<Match> matches(BufferedImage target);

        default Option<Match> bestMatch(BufferedImage target) {
            final double bestSimilarity = 0.75;
            return matches(target)
                    .takeUntil(it -> it.similarity() < bestSimilarity)
                    .toList()
                    .sortBy(Match::similarity)
                    .lastOption();
        }
    }

    interface Match {

        double similarity();

        Rectangle region();

        BufferedImage source();

        BufferedImage image();
    }

    Recogniser recognise(BufferedImage image);
}

File: backend/src/main/java/io/chesslave/eyes/Vision.kt
package io.chesslave.eyes

import javaslang.collection.Stream
import javaslang.control.Option
import java.awt.Rectangle
import java.awt.image.BufferedImage

interface Vision {

    interface Recogniser {

        fun match(target: BufferedImage): Option<Match>

        fun matches(target: BufferedImage): Stream<Match>

        fun bestMatch(target: BufferedImage): Option<Match> =
                matches(target).toList().sortBy { it.similarity() }.lastOption()
    }

    interface Match {

        fun similarity(): Double

        fun region(): Rectangle

        fun source(): BufferedImage

        fun image(): BufferedImage
    }

    fun recognise(image: BufferedImage): Recogniser
}
